### PR TITLE
docs(docs-infra): Add cross-links to API pages

### LINKED ISSUE #55571: No more subclasses or `implements`/`extends` information in API doc
### Describe the problem that you experienced

I was jumping back and forth between the old and new doc sites and noticed that the new site contains much less information on a class/abstract class/interface. For example, compare the two pages on `ViewRef`:

- https://angular.dev/api/core/ViewRef

Screenshot 2024-04-28 124526

- https://angular.io/api/core/ViewRef

Screenshot 2024-04-28 124554

So a whole lot is missing: sub-classes, base classes, comments on which methods come from the inheritance, etc. I'd argue the old API doc is much more useful when digging a little deeper and is more friendly for self-exploration. I found it very useful when I first learned Angular and later when I produced tutorials and sessions to teach others about Angular.

### Enter the URL of the topic with the problem

https://angular.dev/api/core/ViewRef

### Describe what you were looking for in the documentation

_No response_

### Describe the actions that led you to experience the problem

_No response_

### Describe what you want to experience that would fix the problem

_No response_

### Add a screenshot if that helps illustrate the problem

_No response_

### If this problem caused an exception or error, please paste it here

_No response_

### If the problem is browser-specific, please specify the device, OS, browser, and version

_No response_

### Provide any additional information here in as much as detail as you can

_No response_

### LINKED ISSUE #57128: docs (docs-infra): provide links to mentioned API references on API reference detail pages
### Describe the problem that you experienced

When visiting an API reference detail page, other mentioned API references (e.g. Interfaces, Classes, etc.) on that page do not have links to take you directly to the details page of that reference.

e.g. Visit https://angular.dev/api/router/Router#navigate and now try to visit the mentioned `NavigationExtras` Interface to see what it looks like. You will have to either go back to the overview page and search for `NavigationExtras` or try to find it in the sidenav between all `Router` API references.

On https://v17.angular.io/api/router/Router#navigate you can just click on `NavigationExtras` and visit the detail page for it directly.

### Enter the URL of the topic with the problem

https://angular.dev/api

### Describe what you were looking for in the documentation

_No response_

### Describe the actions that led you to experience the problem

_No response_

### Describe what you want to experience that would fix the problem

I want all mentioned API references on API detail pages to have a link to their detail page provided, to avoid having to manually search for it.

### Add a screenshot if that helps illustrate the problem

_No response_

### If this problem caused an exception or error, please paste it here

_No response_

### If the problem is browser-specific, please specify the device, OS, browser, and version

_No response_

### Provide any additional information here in as much as detail as you can

_No response_

### LINKED ISSUE #57183: API docs are missing info on generic type parameters
### Describe the problem that you experienced

I want to quickly look up whether a given class has any generic type parameters and if so, what the defaults of those are.

For example, I should be able to quickly see that the class `FormGroup` is actually defined as `FormGroup<TControl = any>`.

However, generics are not currently part of the displayed API docs.  This can be confusing because the documented class members may reference the generic type name (e.g., in this case `TControl`) with no clear indication of where that type name originates.

### Enter the URL of the topic with the problem

https://angular.dev/api/forms/FormGroup

### PR DESCRIPTION
**Pull request by JeanMeche**
This commit adds links to our own symbols in the API pages, including the Code TOC, the description and the Usage notes. 

This commit also add the implement/extends symbols for classes & interfaces. 

The generation process now logs warnings for links that couldn't not be generated because the symbols can be resolved in the given context. 

fixes https://github.com/angular/angular/issues/57128
fixes https://github.com/angular/angular/issues/57183
fixes #55571

Worthy examples: 
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/core/afterRender
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/common/LocationStrategy?tab=description
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/common/NgForOfContext
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/common/HashLocationStrategy

### CODE REVIEW

File: adev/shared-docs/pipeline/api-gen/rendering/processing.ts [hunk 1]
```diff
@@ -39,38 +39,42 @@ import {getTypeAliasRenderable} from './transforms/type-alias-transforms';
 export function getRenderable(
   entry: DocEntry,
   moduleName: string,
+  symbols: SymbolMap,
 ): DocEntryRenderable | CliCommandRenderable {
   if (isClassEntry(entry)) {
```
Comment (JeanMeche) [inline adev/shared-docs/pipeline/api-gen/rendering/processing.ts:44]: Thought: Should `symbols` be a scoped global so we don't need to passed all the way down ?

File: adev/shared-docs/pipeline/api-gen/extraction/index.ts [hunk 2]
```diff
@@ -54,12 +54,20 @@ function main() {
       return result.concat(JSON.parse(readFileSync(path, {encoding: 'utf8'})) as DocEntry[]);
     }, []);
 
-  const extractedEntries = program.getApiDocumentation(entryPointExecRootRelativePath);
+  const apiDoc = program.getApiDocumentation(entryPointExecRootRelativePath);
+  const extractedEntries = apiDoc.entries;
   const combinedEntries = extractedEntries.concat(extraEntries);
 
+  // Add the module symbols also to the map
+  const strippedModuleName = moduleName.replace('@angular/', '');
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/extraction/index.ts:62]: I think this shows (for the other PR with the core/globals) that we need to maybe have `module_name`, and a "title" field? I'm worried we are using `module_name` for more, but then `Dev-mode globals` breaks

File: adev/shared-docs/pipeline/api-gen/rendering/entities.ts [hunk 3]
```diff
@@ -205,6 +205,8 @@ export interface InitializerApiFunctionEntry extends DocEntry {
   };
 }
 
+export type SymbolMap = Map<string, string>;
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/entities.ts:208]: Nit: Add description

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 4]
```diff
@@ -62,6 +66,11 @@ export class DocsExtractor {
         // We want the real source file of the declaration.
         const realSourceFile = node.getSourceFile();
 
+        const importedSymbols = this.getImportedSymbols(realSourceFile);
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:69]: This computation happens for every export. I also wonder, why is this necessary? can you explain? i.e. why do we need to extract the "imported" symbols and returning them as `symbols` here is a bit ambiguous I feel like.

**Comment by devversion** [comment]
I like the general approach, but I also don't know how it was done before in the early versions of angular.dev, or why we stopped/reverted the logic, so I will defer to @jelbourn on this for now.

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts (hunk 4 above)
Comment (JeanMeche) [reply packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:69]: I've added a comment, let me know if it makes it more clear for you.

File: adev/shared-docs/pipeline/api-gen/extraction/index.ts (hunk 2 above)
Comment (JeanMeche) [reply adev/shared-docs/pipeline/api-gen/extraction/index.ts:62]: Yes, we'll be able to reuse the logic that I added in #57365

**Comment by josephperrott** [comment]
Deferring to other reviewers for this specific change.

**Comment by JeanMeche** [comment]
Currently trying to reduce the scope of this PR, as it might feel a bit large. 
I think the important part are the `compiler-cli` changes.

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 5]
```diff
@@ -30,6 +30,8 @@ import {extractTypeAlias} from './type_alias_extractor';
 
 type DeclarationWithExportName = readonly [string, ts.Declaration];
 
+const privateModules = new Set(['core/primitives/signals']);
```
Comment (jelbourn) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:33]: Needs a comment explaining what this is and why it's necessary

(in general, I'm looking for comments to tell me _why_ the code does something over explaining _what_ the code does)

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 6]
```diff
@@ -902,7 +903,7 @@ export class NgCompiler {
    * @param entryPoint Path to the entry point for the package for which API
    *     docs should be extracted.
    */
-  getApiDocumentation(entryPoint: string): DocEntry[] {
+  getApiDocumentation(entryPoint: string): {entries: DocEntry[]; symbols: SymbolMap} {
```
Comment (jelbourn) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts:906]: I'd want to create a full type for this return value with appropriate JsDoc. Just `symbols` by itself, too, is a bit vague. Maybe something like `importedSymbols` or `allImportedSymbols`

File: packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts [hunk 7]
```diff
@@ -343,6 +378,15 @@ class ClassExtractor {
     const modifiers = this.declaration.modifiers ?? [];
     return modifiers.some((mod) => mod.kind === ts.SyntaxKind.AbstractKeyword);
   }
+
+  /**
+   * Check wether a member has a private computed property name like [ɵWRITABLE_SIGNAL]
+   */
+  private hasPrivateComputedProperty(property: PropertyLike) {
```
Comment (jelbourn) [inline packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts:385]: Needs a comment that explains _why_ this is necessary

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 8]
```diff
@@ -136,6 +156,40 @@ export class DocsExtractor {
       ([a, declarationA], [b, declarationB]) => declarationA.pos - declarationB.pos,
     );
   }
+
+  /**
+   * @returns the symbols from other @angular packages
+   */
+  private getImportedSymbols(sourceFile: ts.SourceFile): SymbolMap {
```
Comment (jelbourn) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:163]: I would break this out into its own extractor file

**Review by jelbourn — COMMENTED** [review]
I got part of the way through before I need to run for the day, but here's my initial comments

Not sure if you were planning to do this already once you got an okay on the approach, but I'll mention anyway: we should split this change up into more granular commits. In particular:
* Separating the symbol collection into its own commit (and this also needs tests)
* Changing the representation of class inheritance in its own commit
* Minimum necessary changes to the renderer to accomodate any `DocEntry` structure changes
* Further changes to surface the links in the rendered output

**Comment by JeanMeche** [comment]
I splitted the changes accross multiple commits, I should ease the review of all the changes (or we can spin them off in other PRs)

File: adev/shared-docs/pipeline/api-gen/extraction/index.ts [hunk 9]
```diff
@@ -54,12 +54,20 @@ function main() {
...
 
   const output = JSON.stringify({
     moduleName: moduleName,
     entries: combinedEntries,
+    symbols: [
+      // Imported symbols from other packages
+      ...apiDoc.symbols.entries(),
+
+      // Exported symbols from the current package
+      ...apiDoc.entries.map((entry) => [entry.name, moduleName]),
```
Comment (JeanMeche) [inline adev/shared-docs/pipeline/api-gen/extraction/index.ts:69]: @jelbourn Do you think we should do this concatenation during extraction ?

**Review by jelbourn — APPROVED** [review]
Overall looks good

File: adev/shared-docs/pipeline/api-gen/rendering/transforms/code-transforms.ts [hunk 10]
```diff
@@ -428,3 +435,28 @@ function appendPrefixAndSuffix(entry: DocEntry, codeTocData: CodeTableOfContents
...
   }
 }
+
+/**
+ * Replaces any code block that isn't already wrapped by an anchor element
+ * by a link if the symbol is known
+ */
+export function addApiLinksToHtml(htmlString: string): string {
+  const result = htmlString.replace(
+    /(?<!<a[^>]*>)(<(?:(?:span)|(?:code))[^>]*>\s*)([^<]*?)(\s*<\/(?:span|code)>)/g,
```
[the remark this answers is not in the document]
Comment (JeanMeche) [reply adev/shared-docs/pipeline/api-gen/rendering/transforms/code-transforms.ts:445]: It is valid, make sure that the visualizer supports ES regex, like https://regex101.com/

File: packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+ * @returns a map Symbol => Package, eg: ApplicationRef => @angular/core
+ */
+export function getImportedSymbols(sourceFile: ts.SourceFile): Map<string, string> {
+  const importSpecifiers = new Map<string, string>();
+
+  function visit(node: ts.Node) {
+    if (ts.isImportDeclaration(node)) {
+      let moduleSpecifier = node.moduleSpecifier.getText(sourceFile).replace(/['"]/g, '');
+
+      if (moduleSpecifier.startsWith('@angular')) {
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts:23]:
```suggestion
      if (moduleSpecifier.startsWith('@angular/')) {
```

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts:23]: would it be better/more generic if we filter everything that is not a relative path?

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 12]
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
         // We want the real source file of the declaration.
         const realSourceFile = node.getSourceFile();
 
+        /**
+         * `sourceFile` is generaly a `export * from './public_api';` with no imports.
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:74]:
```
         * The `sourceFile` from `extractAll` is the main entry-point file of a package.
         * Usually following a format like `export * from './public_api';`, simply re-exporting.
```

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 13]
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
...
+         * It is necessary to pick-up every import from the real source files.
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:75]:
```
         * It is necessary to pick-up every import from the actual source files
         * where declarations are living, so that we can determine what symbols
         * are actually referenced in the context of that particular declaration
```

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 14]
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
...
+         * This allows to include symbols from other packages (like @angular/core)
+         * By doing this, the generation remains independant from other packages
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:77]:
```suggestion
         * By doing this, the generation remains independent from other packages
```

File: adev/shared-docs/pipeline/api-gen/rendering/symbol-context.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+export function setSymbols(newSymbols: Map<string, string>): void {
+  symbols = newSymbols;
+}
+
+/**
+ * Returns the module name of a symbol.
+ * eg: 'ApplicationRef' => 'core', 'FormControl' => 'forms'
+ */
+export function getModuleName(symbol: string): string | undefined {
+  return symbols.get(symbol)?.replace('@angular/', '');
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/symbol-context.ts:17]: Should `symbol` in the extractor simply have the "package name" already? It feels like we are spreading all these cutoffs throughout various places.

File: adev/shared-docs/pipeline/api-gen/extraction/index.ts [hunk 16]
```diff
@@ -54,12 +54,20 @@ function main() {
...
-  const extractedEntries = program.getApiDocumentation(entryPointExecRootRelativePath);
+  const apiDoc = program.getApiDocumentation(entryPointExecRootRelativePath);
+  const extractedEntries = apiDoc.entries;
   const combinedEntries = extractedEntries.concat(extraEntries);
 
   const output = JSON.stringify({
     moduleName: moduleName,
     entries: combinedEntries,
+    symbols: [
+      // Imported symbols from other packages
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/extraction/index.ts:65]:
```suggestion
      // Symbols referenced, originating from other packages
```

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 17]
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
...
+         */
+        const importedSymbols = getImportedSymbols(realSourceFile);
+        importedSymbols.forEach((moduleName, symbolName) => {
+          if (privateModules.has(moduleName) || symbolName.startsWith('ɵ')) {
+            return;
+          }
+          symbols.set(symbolName, moduleName);
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:84]: Is this logic 100% correct? Imagine two declarations from different files import a different `ApplicationRef`. The resolved imports would be different, from different packages potentially? 

If so, please also elaborate this in a comment for future-proofness.

**Review by devversion — APPROVED** [review]
LGTM, great work! Minor comments in the compiler

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts [hunk 18]
```diff
@@ -27,9 +27,13 @@ import {
   isInitializerApiFunction,
 } from './initializer_api_function_extractor';
 import {extractTypeAlias} from './type_alias_extractor';
+import {getImportedSymbols} from './import_extractor';
 
 type DeclarationWithExportName = readonly [string, ts.Declaration];
 
+/** The modules here doesn't expose symbols publicly as they are considered private */
+const privateModules = new Set(['core/primitives/signals', 'core/primitives/event-dispatch']);
```
[the remark this answers is not in the document]
Comment (JeanMeche) [reply packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:35]: Understood, let's handle this in a follow-up !

File: packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts (hunk 11 above)
Comment (JeanMeche) [reply packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts:23]: That would also expose symbols from RxJS, which is not wanted.

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts:23]: It might be the responsibility of the "consumer" of the symbols list to filter out items that cannot be rendered. Or even cooler, we could link to the RxJS docs that way 😄

File: adev/shared-docs/pipeline/api-gen/rendering/symbol-context.ts (hunk 15 above)
Comment (JeanMeche) [reply adev/shared-docs/pipeline/api-gen/rendering/symbol-context.ts:17]: With the latest refactors, it's the only place where we do it (we also don't it anymore in #57365)

File: packages/compiler-cli/src/ngtsc/docs/src/extractor.ts (hunk 17 above)
Comment (JeanMeche) [reply packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:84]: This would indeed be an issue. 

I think the only existing case today is for `inject` which is exposed by both `core` and `core/testing` (we had that same issue on angular.io). 

We could either have an array in the map or drop the entry as it is ambiguous.

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:92]: As discussed offline, we will be erroring if we find conflicts. Right now this shouldn't be the case, especially since we are better than AIO with that, given that "imports are detected via the local context of declarations". e.g. core/testing `inject` will be properly linked in test entry-points, while for non-testing, `core` `inject` may be used.

**Comment by AndrewKushnir** [comment]
@JeanMeche there is a conflict with the `main` branch. Could you please rebase and resolve the merge conflict when you get a chance? Thank you.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:668991633:BASE:668991673:1724953829423:833a5110).

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 01030d58a7162cda2fa597ac77e29ea7066b1a50.

The changes were merged into the following branches: main, 18.2.x
